tests/helpers.py

```diff
@@ -17,34 +17,48 @@ def run_subprocess(command, working_dir='.', env=None, expected_returncode=0, in
     Helper function to run a shell command and report success/failure
     depending on the exit status of the shell command.
     """
+    env_ = os.environ.copy()
     if env is not None:
-        env_ = os.environ.copy()
         env_.update(env)
-        env = env_
+    env = env_
+
+    # Detect if we're running in CircleCI.
+    # https://circleci.com/docs/2.0/env-vars/
+    circleci = 'CIRCLECI' in env
+
+    # CircleCI has its own timeout mechanism. Disable ours.
+    timeout = None if circleci else 10*60
 
     # Note we need to capture stdout/stderr from the subprocess,
     # then print it, which pytest will then capture and
     # buffer appropriately
     print(working_dir + " > " + " ".join(command))
-    proc = subprocess.Popen(
-        command,
-        stdin=subprocess.PIPE,
-        stdout=subprocess.PIPE,
-        stderr=subprocess.STDOUT,
-        cwd=working_dir,
-        env=env,
-    )
-
-    try:
-        out, _ = proc.communicate(input=input, timeout=480)
-    except subprocess.TimeoutExpired:
-        proc.kill()
-        out, _ = proc.communicate()
-
-    if not(ignore_returncode) and (proc.returncode != expected_returncode):
-        print(out.decode('utf-8'))
-        assert False, "Got unexpected return code {}".format(proc.returncode)
-    return out.decode('utf-8')
+
+    while 1: # We might need to re-run the command
+        proc = subprocess.Popen(
+            command,
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+            cwd=working_dir,
+            env=env,
+        )
+
+        try:
+            out, _ = proc.communicate(input=input, timeout=timeout)
+        except subprocess.TimeoutExpired as T:
+            proc.kill()
+            assert False, str(T)
+
+        # CircleCI is welcome to kill this python process, otherwise
+        # we'll keep trying the subprocess.
+        if circleci and proc.returncode == -9:
+            continue
```

Comment:
Given this is an endless loop, would it make sense to do an output if we hit this condition such that we're a bit more informed what's going on when reading the logs?

Reply:
The infinite loop is not necessary at all, since the test runs of this PR revealed that CircleCI was not sending the SIGKILLs. I'll remove it.